src/internal/m365/onedrive/permission.go

```diff
@@ -303,26 +354,49 @@ func UpdateLinkShares(
 		}
 		lsbody.SetAdditionalData(ad)
 
-		if first {
+		if !didReset {
 			// The only way to delete any is to use this and so if
 			// we have any deleted items, we can be sure that all the
 			// inherited permissions would have been removed.
 			lsbody.SetRetainInheritedPermissions(ptr.To(len(lsRemoved) == 0))
 
 			// This value only effective on the first call, but lets
 			// make sure to not send it on followups.
-			first = false
+			didReset = true
 		}
 
 		newLS, err := upils.PostItemLinkShareUpdate(ictx, driveID, itemID, lsbody)
 		if err != nil {
-			return clues.Stack(err)
+			return didReset, clues.Stack(err)
 		}
 
 		oldLinkShareIDToNewID[ls.ID] = ptr.Val(newLS.GetId())
 	}
 
-	return nil
+	// It is possible to have empty link shares even though we should
+	// have inherited one if the user creates a link using
+	// `retainInheritedPermissions` as false, but then deleted it.  We
+	// can recreate this by creating an anonymous one and deleting it.
+	if len(lsRemoved) > 0 && len(lsAdded) == 0 {
+		lsbody := drives.NewItemItemsItemCreateLinkPostRequestBody()
+		lsbody.SetType(ptr.To("view"))
+		lsbody.SetScope(ptr.To("anonymous"))
```

Comment:
I think we need to be really careful with this. For example, is it creating a link share that anyone (logged in or not) can access? It's unlikely that anyone could guess the URL, but it's still a security leak if anyone can see it. Plus, there's no way we can really bound the time the link would actually be there for because it's possible corso could crash between making the link and deleting it. Or it could even just fail to delete the link for some reason. If we can't create a link share that nobody can access then maybe we should just add this to the known issues list

Or is it possible to create a link share that only the owner of the file can access? If that's difficult or involves a lot of code add that support in another PR